aiida_aurora/utils/analyzers.py

```diff
@@ -0,0 +1,193 @@
+    """
+
+    logger = AIIDA_LOGGER.getChild("monitor")
+
+    def set_logger(self, logger: LoggerAdapter) -> None:
+        """Set the analyzer logger.
+
+        Parameters
+        ----------
+        `logger` : `LoggerAdapter`
+            The logger of the analyzed calculation node.
+        """
+        self.logger = logger
+
+    def analyze(self, snapshot: dict) -> Optional[str]:
+        """Analyze the experiment snapshot against a condition.
+
+        Condition is defined in subclass analyzers.
+
+        Parameters
+        ----------
+        `snapshot` : `dict`
+            The loaded snapshot dictionary.
+
+        Returns
+        -------
+        `Optional[str]`
+            A string if a defined condition has been met,
+            `None` otherwise.
+        """
+        raise NotImplementedError
+
+
+class CapacityAnalyzer(Analyzer):
+    """A battery capacity analyzer.
+
+    Attributes
+    ==========
+    `check_type` : `str`
+        The half-cycle to analyze (charge/discharge),
+        `"discharge_capacity"` by default.
+    `threshold` : `float`
+        The capacity threshold in percent, `0.8` by default.
+    `consecutive_cycles` : `int`
+        The number of required below-threshold consecutive cycles,
+        `2` by default
+    """
+
+    def __init__(
+        self,
+        check_type="discharge_capacity",
+        threshold=0.8,
+        consecutive_cycles=2,
+    ) -> None:
+        """`CapacityAnalyzer` constructor.
+
+        Parameters
+        ----------
+        `check_type` : `str`
+            The half-cycle to analyze,
+            `"discharge_capacity"` by default.
+        `threshold` : `float`
+            The capacity threshold in percent, `0.8` by default.
+        `consecutive_cycles` : `int`
+            The number of required consecutive cycles,
+            `2` by default.
+
+        Raises
+        ------
+        `TypeError`
+            If `check_type` is not supported.
+        """
+
+        if check_type not in {"discharge_capacity", "charge_capacity"}:
+            raise TypeError(f"{check_type=} not supported")
+
+        self.threshold = threshold
+        self.consecutive = consecutive_cycles
+        self.is_discharge = check_type == "discharge_capacity"
+
+    def analyze(self, snapshot: dict) -> Optional[str]:
+        """Analyze the snapshot.
+
+        Check if capacity has fallen below threshold for required
+        consecutive cycles.
+
+        Parameters
+        ----------
+        `snapshot` : `dict`
+            The loaded snapshot dictionary.
+
+        Returns
+        -------
+        `Optional[str]`
+            If condition is met, an exit message, `None` otherwise.
+        """
+
+        self.capacities = self._get_capacities(snapshot)
+
+        self.cycles = len(self.capacities)
+
+        if self.cycles < 1:
+            return None
+
+        self.current_capacity = self.capacities[-1]
+        self.threshold = self.threshold * self.capacities[0]
```

Comment:
Does the `CapacityAnalyzer` get instantiated every time it has to perform the analysis or the instance gets re-used? Because if `CapacityAnalyzer.analyze` gets ran more than once per object it will start accumulating products with `self.capacities[0]`, right? I would say even if now we just re-create it each time, it would be better to implement this differently to avoid side-effects.